File: Design-Patterns-Using-Ruby/State/StatePattern2.rb
# Define State interface
module State
    def handle(context)
        raise NotImplementedError, 'Subclasses must implement the handle method'
    end
end

# Define Context class
class Context
    attr_accessor :current_state

    def initialize(state)
        @current_state = state
    end

    def change_state(state)
        @current_state = state
    end

    def request
        @current_state.handle(self)
    end
end

# Define ConcreteState1 class implementing State
class ConcreteState1
    include State

    def handle(context)
        puts 'ConcreteState1 handle'
        context.change_state(ConcreteState2.new)
    end
end

# Define ConcreteState2 class implementing State
class ConcreteState2
    include State

    def handle(context)
        puts 'ConcreteState2 handle'
        context.change_state(ConcreteState1.new)
    end
end

# Client code
state1 = ConcreteState1.new
context = Context.new(state1)
context.request
context.request

=begin 
ConcreteState1 handle
ConcreteState2 handle
=end

File: Design-Patterns-Using-Go/State/StatePattern2.go
package main

import "fmt"

type Context struct {
	currentState State
}

func NewContext(state State) *Context {
	return &Context{
		currentState: state,
	}
}

func (c *Context) changeState(state State) {
	c.currentState = state
}

func (c *Context) request() {
	c.currentState.handle(c)
}

type State interface {
	handle(context *Context)
}

type ConcreteState1 struct{}

func (cs1 *ConcreteState1) handle(context *Context) {
	fmt.Println("ConcreteState1 handle")
	context.changeState(&ConcreteState2{})
}

type ConcreteState2 struct{}

func (cs2 *ConcreteState2) handle(context *Context) {
	fmt.Println("ConcreteState2 handle")
	context.changeState(&ConcreteState1{})
}

// Client code
func main() {
	state1 := &ConcreteState1{}
	context := NewContext(state1)
	context.request()
	context.request()
}

/*
ConcreteState1 handle
ConcreteState2 handle
*/